spices/SPICE-0013-bytes-standard-library.adoc

```diff
@@ -0,0 +1,376 @@
+
+==== `FileOutput` class
+
+A new property is added to the `FileOutput` class, called `bytes`.
+It defaults to the UTF-8 bytes of `text`.
+
+[source,pkl]
+----
+class FileOutput {
+  /* etc */
+
+  /// The textual rendered output.
+  text: String = renderer.renderDocument(value)
+
+  /// The underlying byte array of [text], in UTF-8 encoding.
+  @Since { version = "0.29.0" }
+  bytes: Bytes = text.toBytes()
+}
+----
+
+=== Renderer changes
+
+Some renderers will be able to render `Bytes` out of the box. These are:
+
+* PcfRenderer
+* PlistRenderer
+
+The following Pkl:
+
+[source,pkl]
+----
+res = Bytes("AQIDBA==")
+----
+
+Produces the following PCF:
+
+[source,pkl]
+----
+res = Bytes("AQIDBA==")
+----
+
+Produces the following Plist:
+
+[source,xml]
+----
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+  <key>res</key>
+  <data>AQIDBA==</data>
+</dict>
+</plist>
+----
+
+All other renderers will require a converter to be defined for `Bytes`, and throw an error otherwise.
+
+=== Pkl Binary Encoding changes
+
+When encoded to https://pkl-lang.org/main/current/bindings-specification/binary-encoding.html[Pkl Binary Encoding], `Bytes` will be encoded as an array of two elements.
+
+The first element is `0xF`.
+
+The second element is the binary contents of `Bytes`.
+
+=== Codegen changes
+
+Code generators should convert `bytes` into the respective type in the target language.
+
+These will be the mappings for the libraries maintained by the core team:
+
+|===
+|Language |Type
+
+|Java
+|`byte[]`
+
+|Kotlin
+|`ByteArray`
+
+|Go
+|`[]byte`
+
+|Swift
+|`[UInt8]`
+|===
+
+=== Changes to `pkl eval`
+
+By default, `pkl eval` will change to evaluate `output.bytes` of a module.
+Additionally, `pkl eval -m` will change to evaluate the `output.files[*].bytes` of each file output.
+
+These raw bytes will be written either to standard out, or to the specified output file.
+
+This change means that it is possible for Pkl to write any content, not just UTF-8 encoded strings.
+
+=== Changes to `org.pkl.core.Evaluator`
+
+The `evaluateExpression` method will map Pkl type `Byte` to `byte[]` in Java.
+
+A new method will be added, called `evaluateOutputBytes`, which evaluates the `output.bytes` of a module.
+
+[source,java]
+----
+interface Evaluator {
+  /**
+   * Evaluates a module's {@code output.bytes} property.
+   */
+  byte[] evaluateOutputBytes(ModuleSource moduleSource);
+}
+----
+
+=== Examples
+
+==== Creating `Bytes`
+
+[source,pkl]
+----
+// From a list
+bytes1 = Bytes(List(1, 2, 3, 4))
+
+// From a base64-encoded string
+bytes2 = Bytes("AQIDBA==")
+----
+
+==== Concatenating two `Bytes`
+
+[source,pkl]
+----
+bytes3: Bytes = bytes1 + bytes2
+----
```

Comment:
What about subscript operations? I'd generally expect `bytes1[0]` to return the first byte in the buffer

Reply:
The sentiment of this design is to provide a way to view into a byte array, not really to manipulate them. That's also why this class is called `Bytes`, rather than `ByteList` or something like that.

If you need to manipulate them, you can convert to list, before going back to `Bytes`.

We can explore making `List<UInt8>` more efficient for this.

Reply:
As called out here, this functionality exists and is present in the implementation PR: https://github.com/apple/pkl/pull/1019#discussion_r1991869339

> If you need to manipulate them, you can convert to list, before going back to Bytes.

If the desire is to guide users towards this behaviour, should we disable this functionality in `Bytes` or leave it as it is? I'm less concerned with whether subscript operations should/shouldn't exist and what the supported interface looks like for `Bytes` going forwards.

Reply:
Right; I will get rid of that behavior in accordance with this SPICE, as long as this behavior is agreed upon.

Reply:
Update here: I've updated this SPICE to make `Bytes` itself a sequences of `UInt8`, and added subscript (see [comment](https://github.com/apple/pkl-evolution/pull/14#issuecomment-2870818716))

The actual changes made can be seen in this commit: https://github.com/apple/pkl-evolution/pull/14/commits/2c712eae4eb8d7220d61179f0a8bf85fce6cdb87